# Import libraries

In [1]:
import keras
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical
from keras.layers.convolutional import Conv2D # to add convolutional layers
from keras.layers.convolutional import MaxPooling2D # to add pooling layers
from keras.layers import Flatten # to flatten data for fully connected layers
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import numpy as np
import pandas as pd

C:\Users\Hamad\anaconda3\lib\site-packages\scipy\__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.24.2
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


# Build the neural network

- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error  as the loss function.

In [2]:
def NN():
    
    model = Sequential()
    
    model.add(Dense(units = 10, activation = 'relu'))
    model.add(Dense(units = 1, activation = 'linear'))
    
    model.compile(optimizer='adam', loss='mean_squared_error')
    
    return model

# Read the data

In [3]:
df = pd.read_csv("concrete_data.csv")
df.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

# Split the data into X and Y

In [4]:
X = df.drop('Strength', axis = 1)
y = df['Strength']

print(X.shape)
print(y.shape)

(1030, 8)
(1030,)


# Normalize the data

- Use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

In [5]:
X_norm = (X - X.mean()) / X.std()
X_norm.describe()

Cement  Blast Furnace Slag       Fly Ash         Water  \
count  1.030000e+03        1.030000e+03  1.030000e+03  1.030000e+03   
mean   2.432224e-15       -8.513686e-16  3.837815e-16  1.846743e-15   
std    1.000000e+00        1.000000e+00  1.000000e+00  1.000000e+00   
min   -1.714421e+00       -8.564718e-01 -8.467326e-01 -2.798851e+00   
25%   -8.496407e-01       -8.564718e-01 -8.467326e-01 -7.805147e-01   
50%   -7.911350e-02       -6.014861e-01 -8.467326e-01  1.607513e-01   
75%    6.586406e-01        8.003558e-01  1.001791e+00  4.885554e-01   
max    2.476712e+00        3.309068e+00  2.279976e+00  3.064159e+00   

       Superplasticizer  Coarse Aggregate  Fine Aggregate           Age  
count      1.030000e+03      1.030000e+03    1.030000e+03  1.030000e+03  
mean      -9.641155e-16      6.818710e-15    1.232571e-14  3.640022e-16  
std        1.000000e+00      1.000000e+00    1.000000e+00  1.000000e+00  
min       -1.038638e+00     -2.211064e+00   -2.239829e+00 -7.070160e-01  
25%       -1.038638e+00     -5.262618e-01   -5.317114e-01 -6.120340e-01  
50%        3.269920e-02     -6.326279e-02    7.383152e-02 -2.795973e-01  
75%        6.688058e-01      7.264077e-01    6.288606e-01  1.636517e-01  
max        4.351528e+00      2.213149e+00    2.731735e+00  5.055221e+00

# 1. Split the data into training and testing

- Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_splithelper function from Scikit-learn.

In [6]:
def data_split(X, y):
    return train_test_split(X, y, test_size=0.3)

X_train, X_test, y_train, y_test = data_split(X_norm, y)

# 2. Train the model

- Train the model on the training data using 100 epochs.

In [7]:
def model_training(X_train, y_train, verbose = 0):
    model = NN()
    model.fit(X_train, y_train, epochs = 100, verbose = verbose)
    return model

model = model_training(X_train, y_train, 1)

Epoch 1/100
23/23 [==============================] - 1s 1ms/step - loss: 1602.7351
Epoch 2/100
23/23 [==============================] - 0s 1ms/step - loss: 1587.2812
Epoch 3/100
23/23 [==============================] - 0s 1ms/step - loss: 1572.1476
Epoch 4/100
23/23 [==============================] - 0s 999us/step - loss: 1557.0566
Epoch 5/100
23/23 [==============================] - 0s 1ms/step - loss: 1541.6428
Epoch 6/100
23/23 [==============================] - 0s 970us/step - loss: 1525.8602
Epoch 7/100
23/23 [==============================] - 0s 1ms/step - loss: 1509.5376
Epoch 8/100
23/23 [==============================] - 0s 1ms/step - loss: 1492.3927
Epoch 9/100
23/23 [==============================] - 0s 1ms/step - loss: 1474.4287
Epoch 10/100
23/23 [==============================] - 0s 1ms/step - loss: 1455.6488
Epoch 11/100
23/23 [==============================] - 0s 1ms/step - loss: 1435.5514
Epoch 12/100
23/23 [==============================] - 0s 1ms/step - loss: 1414.56

23/23 [==============================] - 0s 1ms/step - loss: 160.9727
Epoch 100/100
23/23 [==============================] - 0s 2ms/step - loss: 159.7188


# 3. Evaluate the model

- Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

In [8]:
def prediction(model, X_test, y_test):
    y_pred = model.predict(X_test)
    return mean_squared_error(y_test, y_pred)

print(f'MSE : {prediction(model, X_test, y_test)}')

10/10 [==============================] - 0s 1ms/step
MSE : 141.48445322706954


# 4. Repeat steps 1 to 3

- Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

In [9]:
MSE_lst = []

for i in range (50):
    X_train, X_test, y_train, y_test = data_split(X_norm, y)
    print(f'Model {i+1} :-')
    model = model_training(X_train, y_train)
    mse = prediction(model, X_test, y_test)
    MSE_lst.append(mse)

Model 1 :-
10/10 [==============================] - 0s 2ms/step
Model 2 :-
10/10 [==============================] - 0s 117us/step
Model 3 :-
10/10 [==============================] - 0s 0s/step
Model 4 :-
10/10 [==============================] - 0s 1ms/step
Model 5 :-
10/10 [==============================] - 0s 1ms/step
Model 6 :-
10/10 [==============================] - 0s 740us/step
Model 7 :-
10/10 [==============================] - 0s 1ms/step
Model 8 :-
10/10 [==============================] - 0s 0s/step
Model 9 :-
10/10 [==============================] - 0s 2ms/step
Model 10 :-
10/10 [==============================] - 0s 2ms/step
Model 11 :-
10/10 [==============================] - 0s 0s/step
Model 12 :-
10/10 [==============================] - 0s 1ms/step
Model 13 :-
10/10 [==============================] - 0s 1ms/step
Model 14 :-
10/10 [==============================] - 0s 1ms/step
Model 15 :-
10/10 [==============================] - 0s 1ms/step
Model 16 :-
10/10 [==============

# Report for MSE

In [10]:
dic = {
    'MSE' : []
}

models = []

for i in range(len(MSE_lst)):
    models.append(i+1)
    dic['MSE'].append(MSE_lst[i])
    
report = pd.DataFrame(dic)
report.index = models
report

MSE
1   158.238384
2   154.043301
3   196.170561
4   138.437919
5   157.632227
6   166.032033
7   196.991326
8   147.480725
9   161.126104
10  149.923031
11  203.386159
12  224.935889
13  157.065308
14  165.973903
15  157.710856
16  157.299062
17  147.717469
18  171.238387
19  176.103591
20  171.535644
21  167.615294
22  168.099517
23  155.913472
24  172.609828
25  155.327855
26  175.150821
27  144.647325
28  163.491008
29  178.125317
30  165.299600
31  162.492451
32  213.452152
33  171.631053
34  162.225932
35  184.554153
36  153.097067
37  164.060239
38  171.030646
39  143.395749
40  234.583931
41  157.855754
42  179.916756
43  212.003756
44  189.189779
45  178.713933
46  177.937860
47  157.977549
48  146.242376
49  178.384513
50  186.465019

# Report the mean and the standard deviation of the mean squared errors.

In [11]:
print(f"Average of MSE         : {np.mean(MSE_lst)}")
print(f"Standard Deviation MSE : {np.std(MSE_lst)}")

Average of MSEs: 170.6106516231569
Standard Deviation MSEs: 20.651984308130817


# How does the mean of the mean squared errors compare to that from Step B?

The Avg. MSE has reduced from 348.84 to 170.61.
The Standard Deviation in the MSE has reduced from 84.45 to 20.65.

Which make the model is performance is much better than the model in step B.